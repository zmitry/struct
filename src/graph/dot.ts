import { IUndirectedGraph, meta } from './graph';
import { IHierarchy, ROOT_NODE } from './compound-graph';

type Writer = ReturnType<typeof makeWriter>;
export function toDot<N, E>(
  g: IUndirectedGraph<N, E> & IHierarchy,
  args: { intend?: string; compound: true }
): string;

export function toDot<N, E>(
  g: IUndirectedGraph<N, E>,
  args: { intend?: string; compound?: false }
): string;

export function toDot<N, E>(
  g: any,
  { intend = ' ', compound }: { intend?: string; compound?: boolean }
) {
  const graph: IUndirectedGraph<N, E> & IHierarchy = g;
  const writer = makeWriter(intend);
  const directed = g[meta]?.directed;
  const ec = directed ? '->' : '--';
  writer.writeBlock(directed ? 'digraph' : 'graph', () => {
    graph.nodes().forEach(([key, value]) => {
      writer.writeLine(`${id(key)}${attrsToString(value)}`);
    });

    graph.edges().forEach(edge => {
      writer.writeLine(
        `${id(edge.from)}${ec}${id(edge.to)}${attrsToString(edge.value)}`
      );
    });
    if (compound) {
      writeSubgraph(graph, ROOT_NODE, writer);
    }
  });
  return writer.toString();
}
function writeSubgraph(
  g: IHierarchy & IUndirectedGraph<any, any>,
  v: string,
  writer: Writer
) {
  g.getChildren(v).forEach(w => {
    if (g.getChildren(w).length > 0) {
      writer.writeBlock(`subgraph ${id(w)}`, () => {
        writeSubgraph(g, w, writer);
      });
    } else if (v !== ROOT_NODE) {
      writer.writeLine(`${id(w)};'`);
    }
  });
}

function id(obj: any) {
  if (!obj) {
    return '1';
  }
  if (typeof obj === 'number') {
    return obj;
  }

  return `"${String(obj).replace(/"/g, '\\"')}"`;
}

function attrsToString(attrs?: Object) {
  if (attrs && isObject(attrs)) {
    const attrStrs = Object.entries(attrs).map(
      ([key, val]) => `${id(key)}=${id(val)}`
    );
    if (attrStrs.length) {
      return ` [${attrStrs.join(',')}]`;
    }
  }
  return '';
}

function makeWriter(INDENT = '') {
  let indent = '';
  let content = '';
  let shouldIndent = true;
  const write = (str: string) => {
    if (shouldIndent) {
      shouldIndent = false;
      content += indent;
    }
    content += str;
  };
  function indentfn() {
    indent += INDENT;
  }
  function unindent() {
    indent = indent.slice(INDENT.length);
  }
  function writeLine(line = '') {
    write(`${line}\n`);
    shouldIndent = true;
  }
  return {
    indent: indentfn,
    unindent,
    toString() {
      return content;
    },
    writeLine,
    writeBlock(str: string, fn: () => void) {
      writeLine(str + ' {');
      indentfn();
      fn();
      unindent();
      writeLine('}');
    },
    write: write,
  };
}

function isObject(value: unknown): boolean {
  const type = typeof value;
  return value != null && (type === 'object' || type === 'function');
}
