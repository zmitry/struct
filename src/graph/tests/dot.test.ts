import { toDot } from '../dot';
import { createGraph, mapToCompoundGraph } from '../index';

function setup() {
  const digraph = createGraph({
    directed: true,
    events: {},
  });
  const g = mapToCompoundGraph(digraph);
  g.setNode('1');
  g.setNode('2', { color: 'red' });
  g.setEdge('1', '2');
  g.setEdge('3', '4');
  g.setParent('4', '3');

  return {
    graph: g,
  };
}
describe('dot converted', () => {
  it('compound', () => {
    const { graph } = setup();
    expect(toDot(graph, { intend: '  ', compound: true })).toMatchSnapshot();
  });

  it('basic', () => {
    const { graph } = setup();
    expect(toDot(graph, { intend: '  ' })).toMatchSnapshot();
  });
});
