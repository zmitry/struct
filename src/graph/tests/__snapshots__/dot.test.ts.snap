// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dot converted basic 1`] = `
"graph {
  \\"1\\"
  \\"2\\" [\\"color\\"=\\"red\\"]
  \\"3\\"
  \\"4\\"
  \\"1\\"--\\"2\\"
  \\"3\\"--\\"4\\"
}
"
`;

exports[`dot converted compound 1`] = `
"digraph {
  \\"1\\"
  \\"2\\" [\\"color\\"=\\"red\\"]
  \\"3\\"
  \\"4\\"
  \\"1\\"->\\"2\\"
  \\"3\\"->\\"4\\"
  subgraph \\"3\\" {
    \\"4\\";'
  }
}
"
`;
